circuits/cpp/src/aztec3/circuits/kernel/private/native_private_kernel_circuit_ordering.cpp

```diff
@@ -0,0 +1,75 @@
+#include "common.hpp"
+#include "init.hpp"
+
+#include "aztec3/circuits/abis/kernel_circuit_public_inputs.hpp"
+#include "aztec3/circuits/abis/private_kernel/private_kernel_inputs_inner.hpp"
+#include "aztec3/constants.hpp"
+#include "aztec3/utils/dummy_composer.hpp"
+
+#include <cstddef>
+
+namespace aztec3::circuits::kernel::private_kernel {
+
+using aztec3::circuits::abis::KernelCircuitPublicInputs;
+using aztec3::circuits::abis::private_kernel::PrivateKernelInputsInner;
+
+using DummyComposer = aztec3::utils::DummyComposer;
+using CircuitErrorCode = aztec3::utils::CircuitErrorCode;
+
+void chop_pending_commitments(DummyComposer& composer,
+                              std::array<NT::fr, READ_REQUESTS_LENGTH> const& read_requests,
+                              std::array<MembershipWitness<NT, PRIVATE_DATA_TREE_HEIGHT>, READ_REQUESTS_LENGTH> const&
+                                  read_request_membership_witnesses,
+                              std::array<NT::fr, KERNEL_NEW_COMMITMENTS_LENGTH>& new_commitments)
+{
+    // chop commitments from the previous call(s)
+    for (size_t i = 0; i < READ_REQUESTS_LENGTH; i++) {
+        const auto& read_request = read_requests[i];
+        const auto is_transient_read = (read_request_membership_witnesses[i].leaf_index == NT::fr(-1));
+
+        if (is_transient_read) {
+            size_t match_pos = KERNEL_NEW_COMMITMENTS_LENGTH;
+            for (size_t j = 0; j < KERNEL_NEW_COMMITMENTS_LENGTH; j++) {
+                match_pos = (read_request == new_commitments[j]) ? j : match_pos;
+            }
+
+            // chop the pending commitment, i.e., replacing with 0.
+            if (match_pos != KERNEL_NEW_COMMITMENTS_LENGTH) {
+                new_commitments[match_pos] = fr(0);
+            } else {
+                composer.do_assert(
+                    false,
+                    format("transient read request at position [", i, "] does not match any new commitment"),
```

Comment:
Read request may not match any commitment and that should be fine, right?

Reply:
For a transient one, this should not happen, right?
For a non-transient one, I agree.

Reply:
If this does not match, the read request might point to a commitment which might already exist in the data tree and for which no witnesses were supplied.

Reply:
Makes sense, thank you!